theano/sandbox/cuda/tests/test_dnn.py

```diff
@@ -715,6 +715,119 @@ def test_log_softmax2(self):
         utt.assert_allclose(f(inp), f_ref(inp))
 
 
+def test_batchnorm_train():
+    if not cuda.dnn.dnn_available():
+        raise SkipTest(cuda.dnn.dnn_available.msg)
+    if cuda.dnn.version() < (5000, 5000):
+        raise SkipTest("batch normalization requires cudnn v5+")
+    utt.seed_rng()
+
+    for mode in ('per-activation', 'spatial'):
+        for vartype in (T.tensor4, T.tensor3, T.matrix, T.vector):
+            x, scale, bias = (vartype(n) for n in ('x', 'scale', 'bias'))
+            ndim = x.ndim
+            eps = 5e-3  # some non-standard value to test if it's used
+
+            # forward pass
+            out, x_mean, x_invstd = cuda.dnn.dnn_batch_normalization_train(
+                x, scale, bias, mode, eps)
+            # reference forward pass
+            if mode == 'per-activation':
+                axes = (0,)
+            elif mode == 'spatial':
+                axes = (0,) + tuple(range(2, ndim))
+            x_mean2 = x.mean(axis=axes, keepdims=True)
+            x_invstd2 = T.inv(T.sqrt(x.var(axis=axes, keepdims=True) + eps))
+            scale2 = T.addbroadcast(scale, *axes)
+            bias2 = T.addbroadcast(bias, *axes)
+            out2 = (x - x_mean2) * (scale2 * x_invstd2) + bias2
+            # backward pass
+            dy = vartype('dy')
+            grads = T.grad(None, wrt=[x, scale, bias], known_grads={out: dy})
+            # reference backward pass
+            grads2 = T.grad(None, wrt=[x, scale, bias], known_grads={out2: dy})
+            # compile
+            f = theano.function([x, scale, bias, dy],
+                                [out, x_mean, x_invstd, out2, x_mean2, x_invstd2] +
+                                grads + grads2, mode=mode_with_gpu)
+            # run
+            floatX = theano.config.floatX
+            for data_shape in ((10, 20, 30, 40), (4, 3, 1, 1), (1, 1, 5, 5)):
+                data_shape = data_shape[:ndim]
+                param_shape = tuple(1 if d in axes else s
+                                    for d, s in enumerate(data_shape))
+                X = 4 + 3 * numpy.random.randn(*data_shape).astype(floatX)
+                Dy = -1 + 2 * numpy.random.randn(*data_shape).astype(floatX)
+                Scale = numpy.random.randn(*param_shape).astype(floatX)
+                Bias = numpy.random.randn(*param_shape).astype(floatX)
+                outputs = f(X, Scale, Bias, Dy)
+                # compare outputs
+                utt.assert_allclose(outputs[0], outputs[0 + 3])  # out
+                utt.assert_allclose(outputs[1], outputs[1 + 3])  # mean
+                utt.assert_allclose(outputs[2], outputs[2 + 3])  # invstd
+                # compare gradients
+                utt.assert_allclose(outputs[6], outputs[6 + 3])  # dx
+                utt.assert_allclose(outputs[7], outputs[7 + 3], rtol=3e-3)  # dscale
+                utt.assert_allclose(outputs[8], outputs[8 + 3])  # dbias
```

Comment:
for the rtol, can you try different cudnn algo? The problems could be that
some aren't deterministic.

On Thu, Jun 16, 2016 at 2:05 PM, Jan Schlüter <notifications@github.com>
wrote:

> In theano/sandbox/cuda/tests/test_dnn.py
> <https://github.com/Theano/Theano/pull/4582#discussion_r67394988>:
>
> > +                data_shape = data_shape[:ndim]
> > +                param_shape = tuple(1 if d in axes else s
> > +                                    for d, s in enumerate(data_shape))
> > +                X = 4 + 3 * numpy.random.randn(*data_shape).astype(floatX)
> > +                Dy = -1 + 2 * numpy.random.randn(*data_shape).astype(floatX)
> > +                Scale = numpy.random.randn(*param_shape).astype(floatX)
> > +                Bias = numpy.random.randn(*param_shape).astype(floatX)
> > +                outputs = f(X, Scale, Bias, Dy)
> > +                # compare outputs
> > +                utt.assert_allclose(outputs[0], outputs[0 + 3])  # out
> > +                utt.assert_allclose(outputs[1], outputs[1 + 3])  # mean
> > +                utt.assert_allclose(outputs[2], outputs[2 + 3])  # invstd
> > +                # compare gradients
> > +                utt.assert_allclose(outputs[6], outputs[6 + 3])  # dx
> > +                utt.assert_allclose(outputs[7], outputs[7 + 3], rtol=3e-3)  # dscale
> > +                utt.assert_allclose(outputs[8], outputs[8 + 3])  # dbias
>
> This high rtol is actually a bit troubling. Everything else is very close
> to what we get with vanilla Theano, but the gradient wrt. scale is a bit
> off. For example, we get (cuDNN vs reference):
>
> [[[[ 504.36730957]]    [[[[ 504.3671875 ]]
>   [[-286.85342407]]      [[-286.85232544]]
>   [[-163.40234375]]      [[-163.40229797]]
>   [[  21.92313004]]      [[  21.92422295]]
>   [[ 226.3973999 ]]      [[ 226.39941406]]
>   [[  -0.51444209]]      [[  -0.51332498]]
>   [[  42.07894897]]      [[  42.08079529]]
>   [[  26.67364693]]      [[  26.6755085 ]]
>   [[ 154.5411377 ]]      [[ 154.54122925]]
>   [[  45.03813934]]      [[  45.03807831]]
>   [[  84.71562958]]      [[  84.71569061]]
>   [[ -62.2019043 ]]      [[ -62.20304108]]
>   [[ 259.81713867]]      [[ 259.81518555]]
>   [[  22.97838593]]      [[  22.97837639]]
>   [[ 141.40872192]]      [[ 141.41171265]]
>   [[ 191.33883667]]      [[ 191.33883667]]
>   [[  22.5411129 ]]      [[  22.5394516 ]]
>   [[ 211.42269897]]      [[ 211.42268372]]
>   [[-179.14613342]]      [[-179.14613342]]
>   [[ 103.85586548]]]]    [[ 103.85778809]]]]
>
> Some values are on spot (the second to last), some are far off in relative
> terms (-0.513 vs -0.514). Not sure if this is just different rounding or
> something wrong with cuDNN. In any case, it's probably close enough for
> learning.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/4582/files/acf5edee8b337f37d660f969841b3bf8a2a3bb65#r67394988>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AALC-9leZq2fRYvRG5V557rmqOA6JLSlks5qMZBrgaJpZM4Ituzu>
> .
>